openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -195,14 +197,34 @@ def _verify_device_uuid(self):
 
     def upgrade(self, image):
         self._test_connection()
+        self._check_cancellation()
         self._verify_device_uuid()
+        self._check_cancellation()
         checksum = self._test_checksum(image)
+        self._check_cancellation()
         remote_path = self.get_remote_path(image)
-        self.upload(image.file, remote_path)
+        self.upload(image, remote_path)
+        self._check_cancellation()
         self._test_image(remote_path)
         self._reflash(remote_path)
         self._write_checksum(checksum)
 
+    def _check_cancellation(self):
+        """
+        Check if the upgrade operation has been cancelled.
+        """
+        self.upgrade_operation.refresh_from_db()
+
+        if self.upgrade_operation.status == "cancelled":
+            self.log(_("Upgrade operation has been cancelled by user"))
+
+            if self._non_critical_services_stopped:
+                self.log(_("Restarting non-critical services..."))
+                self._start_non_critical_services()
```

Comment:
so, this method, which is the one in charge of basically stopping the upgrade, already restarts the services, right?

So what is `_restart_services_after_cancellation` needed for? It seems redundant to me. If it's not redundant, please explain.